### ISSUE TITLE
ceedling exit code is 0 on unit test failure

### ISSUE THREAD

**Issue by ekohandel** [ThrowTheSwitch/Ceedling#174]
When running `ceedling test:all` with failing unit tests, the process returns a 0 status code. It should return 1 as the status code so CI tools can detect the failure.

I am using the following versions of the farmework:

>Ceedling:: 0.27.0
CException:: 1.3.1.18
CMock:: 2.4.4.215
Unity:: 2.4.1.120

This issue was reported in #117 and was reported to be fixed in #130; however, I seem to still experience this problem.

I have setup a repository that reproduces this problem. It can be found at: https://github.com/nmisoft/ceedling_exit_code_bug

Interestingly using `rake test:all` will correctly return 1 as the status code.

I have reproduced this problem both on Windows and Linux.

## Windows
`OS Version: 10.0.14393 N/A Build 14393`

#### Using `ceedling`
```
>ceedling test:all
---------------------
BUILD FAILURE SUMMARY
---------------------
Unit test failures.

ERROR: Aborted with Exit Code 1
```

```
>echo %ERRORLEVEL%
0
```

#### Using `rake`
```
>rake test:all
Test 'test_demo.c'
------------------
Running test_demo.out...

-------------------
FAILED TEST SUMMARY
-------------------
[test_demo.c]
  Test: test_demo_fail
  At line (16): "Expected TRUE Was FALSE"

--------------------
OVERALL TEST SUMMARY
--------------------
TESTED:  1
PASSED:  0
FAILED:  1
IGNORED: 0

---------------------
BUILD FAILURE SUMMARY
---------------------
Unit test failures.
```

```
>echo %ERRORLEVEL%
1
```

## Linux
`Linux 4.4.0-78-generic #99-Ubuntu SMP Thu Apr 27 15:29:09 UTC 2017 x86_64 GNU/Linux`

#### Using `ceedling`
```
$ ceedling test:all


Test 'test_demo.c'
------------------
Running test_demo.out...

-------------------
FAILED TEST SUMMARY
-------------------
[test_demo.c]
  Test: test_demo_fail
  At line (16): "Expected TRUE Was FALSE"

--------------------
OVERALL TEST SUMMARY
--------------------
TESTED:  1
PASSED:  0
FAILED:  1
IGNORED: 0

---------------------
BUILD FAILURE SUMMARY
---------------------
Unit test failures.
```

```
$ echo $?
0
```

#### Using `rake`
```
$ rake test:all


Test 'test_demo.c'
------------------
Running test_demo.out...

-------------------
FAILED TEST SUMMARY
-------------------
[test_demo.c]
  Test: test_demo_fail
  At line (16): "Expected TRUE Was FALSE"

--------------------
OVERALL TEST SUMMARY
--------------------
TESTED:  1
PASSED:  0
FAILED:  1
IGNORED: 0

---------------------
BUILD FAILURE SUMMARY
---------------------
Unit test failures.
```

```
$ echo $?
1
```

**Comment by mvandervoord** [comment]
Thanks for the detailed report. 

In the future, if there is an existing bug, it is more helpful to us to add details to that one than open another bug for the same issue.  

When you say you're testing with Ceedling 0.27.0, does that mean the latest in main or the actual release gem? The 0.27.0 release is where the problem reappeared, so I'm not surprised if that's the version you have issues with. I'd love to know if the bugfix actually works or not. I've left #130 open until I can verify this. Maybe you can help? Can you update your repository to verify the latest in main to see if it fixes the issue when calling Ceedling directly?

Thanks for your help!

**Comment by dpostorivo** [comment]
This is the same issue as in #148

**Comment by ekohandel** [comment]
@mvandervoord I will reopen the existing issue in the future, sorry about that.

The report above is based on the released gem. If I understand this correctly #130 was merged into master before the 0.27.0 release. I did however check this against the latest master and the same broken behavior still exists.

@dpostorivo seems to have fully captured the problem here. I missed #148 prior to creating this issue which really makes this issue a duplicate.

**Comment by dpostorivo** [comment]
@ekohandel  just to confirm, `rake test:all` works for both of your cases?

**Comment by ekohandel** [comment]
@dpostorivo that is correct. I have changed my CI system to run `rake test:all` until this issue is resolved.

**Comment by mvandervoord** [comment]
@dpostorivo and @ekohandel, can either or both of you try the branch refactor/rake-direct? I've reworked the way the ceedling script in bin starts rake tasks. It is now pulling in rake and running them directly instead of shelling out to the command line to do so (which was the source of the majority of the recent bugs and was just plain inefficient). 

At the moment, I know that typing `ceedling help` returns a list without the descriptions. Do you see any other issues with it?

**Comment by dpostorivo** [comment]
@mvandervoord  

I've noticed both clean and clobber now have an error message associated with them


```
Cleaning build artifacts...
(For large projects, this task may take a long time to complete)

ERROR: Ceedling Failed

```

and

```
Clobbering all generated files...
(For large projects, this task may take a long time to complete)

ERROR: Ceedling Failed
```

There's also a warning generated by both: `vendor/ceedling/lib/ceedling/system_wrapper.rb:25: warning: Insecure world writable dir /usr/local in PATH, mode 040777`

**Comment by ekohandel** [comment]
@mvandervoord 

I am seeing similar results as @dpostorivo. There appears to be something else wrong but for what it is worth, I am still getting 0 as the return code even when the tests fail using `ceedling test:all`. `rake test:all` still correctly returns 1 on test failure and is much more verbose in its output.

```
$ ceedling test:all
---------------------
BUILD FAILURE SUMMARY
---------------------
Unit test failures.

ERROR: Aborted with Exit Code 1

$ echo $?
0
```

```
$ rake test:all


Test 'test_demo.c'
------------------
Running test_demo.out...

-------------------
FAILED TEST SUMMARY
-------------------
[test_demo.c]
  Test: test_demo_fail
  At line (16): "Expected TRUE Was FALSE"

--------------------
OVERALL TEST SUMMARY
--------------------
TESTED:  1
PASSED:  0
FAILED:  1
IGNORED: 0

---------------------
BUILD FAILURE SUMMARY
---------------------
Unit test failures.


$ echo $?
1
```

**Comment by mvandervoord** [comment]
@ekohandel, I believe your situation is that it is still calling the original ceedling from your gem. If you copy the ceedling file from ceedling/bin off github and copy it over the same file in your rubygems directory (often something like C:\ruby\lib\ruby\gems\2.1.5\gems\ceedling-0.27.0\bin), I think you will be seeing something more like what @dpostorivo is seeing.

@dpostorivo, could you try renaming your rakefile in your project to something else (like rakefile.old) and see if that has any effect on the situation? It worries me that it's not reporting the ACTUAL error you're getting, just the fact that there WAS an error. I'm not sure what to make of that.

**Comment by ekohandel** [comment]
@mvandervoord, my bad! I made sure to use the new bin and the exit code is now correctly reported as 1 in the case of a failure and the the output is same as what I would get using rake.

I renamed my rake file to rakefile.old and still get the same results when running clobber. It appears that clobber does actually work as all files are recompiled after I run the clobber command.

**Comment by mvandervoord** [comment]
@ekohandel Thanks for the update! 

I'm not sure why both of you get an error when it clobbers. I'm not sure why I don't see that too. I'll try it out on a few different OS's and see if I can narrow the problem down. 

Thanks for the help!

**Comment by mvandervoord** [comment]
I'm wondering if the clobber itself was returning an error code for attempting to clean a file or folder that was no longer present. Because of that, I've added extra protection to the clean and clobber routines with a rescue block. Can one of you check to see if this improves anything (version 4cf99a5)

**Comment by ekohandel** [comment]
@mvandervoord Unfortunately I still get the same result.

With my limited knowledge of Ruby, Rake, and Ceedling, I did some digging and it looks like the `$exit_code` variable in the `SystemWrapper` class is `nil` and so the call to `@ceedling[:system_wrapper].ruby_success` on line 73 of `lib\ceedling\rakefile.rb` returns `false`.

I am not sure what would be the correct way to go about fixing this or I would submit a pull request.

**Comment by dpostorivo** [comment]
I also tried it last night with the same results. I most likely won't have time till the end of the week to dig deeper into it.

**Comment by mvandervoord** [comment]
Hey gentlemen. I have posted a release 0.28.1. It works for me on multiple platforms. Would you mind checking it on yours?

Thanks again!

**Comment by dpostorivo** [comment]
@mvandervoord it works fine on Ubuntu

**Comment by ekohandel** [comment]
@mvandervoord I have tested this on my Linux Mint setup and Windows 10 setup and it works in both cases!

**Comment by mvandervoord** [comment]
Awesome. Thanks for all the help, both of you! Fixed in release 0.28.1

**Comment by gpenghe** [comment]
Still seeing the same issue with Ceedling 0.28.2:
```
  Ceedling:: 0.28.2
CException:: 1.3.1.18
     CMock:: 2.4.6.217
     Unity:: 2.4.3.122
```

**Comment by rleigh-lumiradx** [comment]
With ceedling 0.28.3 on Windows 10:

```
> ceedling test:all

[...]
-------------------------------------------------------------
!!!!FAILURE TO COMPILE/LINK UNIT TEST FILE!!!!
THE FOLLOWING FILES HAD ASSOCIATED COMPILE/LINK ISSUES, NO COVERAGE RECORDED, TEST RESULTS FORCED TO FAIL.
tests/xxx.c

Note: General test related observations in 'build/artifacts/test/observations.out'

---------------------
OVERALL TEST SUMMARY
---------------------
TESTED:   nnn
PASSED:    nnn
FAILED:      20
IGNORED:    nnn


Running Hook post_build...

>echo %errorlevel%
0
```

It looks like there's been a regression or change in behaviour which has re-introduced this bug.

**Comment by jihbuomj** [comment]
Still having issue with 0.29.1 on WSL and 0.29.1 on ruby:2.7 docker image (debian)

```
> ceedling && echo "Exit code: $?"

Test 'test_LedDriver.c'
-----------------------
Generating runner for test_LedDriver.c...
Compiling test_LedDriver_runner.c...
Compiling test_LedDriver.c...
Compiling LedDriver.c...
Compiling unity.c...
Compiling CException.c...
Compiling cmock.c...
Linking test_LedDriver.out...
Running test_LedDriver.out...
test_LedDriver.c:8:test_Led_off_after_create:PASS
test_LedDriver.c:16:test_should_fail:FAIL: This should fail

-----------------------
2 Tests 1 Failures 0 Ignored 
FAIL
Exit code: 0
```
Version Info:
```
> ceedling version

   Ceedling:: 0.29.1
      Unity:: 2.5.0
      CMock:: 2.5.1
 CException:: 1.3.2
```

**Comment by jupe** [comment]
@jihbuomj @mvandervoord This seems to be still valid issue, right ? Is there any regression test for this if it was already fixed once ? Is there some workaround for CI usage?

**Comment by jihbuomj** [comment]
@jupe 

This is the `.gitlab-ci.yml` file I used:

```
test-app:
  image: <local-image>/ceedling:0.29.1
  script:
    - cd tests
    - mkdir build
    - ceedling test:all | tee output
    - tail -n 1 output | if [ "$(cat)" == "FAIL" ]; then exit 1; fi
```
My ceedling image is based on the official ruby image which is based on Debian.

**Comment by zaq32** [comment]
The change that originally fixed this problem https://github.com/ThrowTheSwitch/Ceedling/pull/175/commits/2cae4e55e201ec2f7583e2927430f7a0d16fe1fa#diff-4bca1dda9388536f3fec12ad3707b5f6R370-R373 has been overwritten the very next day: https://github.com/ThrowTheSwitch/Ceedling/commit/ebc385f6ff0b30fb0a1ed7e237fb5e3cf3d23593#diff-4bca1dda9388536f3fec12ad3707b5f6L370-L373 Needles to say that restoration of the very last line fixes the problem for me.

**Comment by rleigh-lumiradx** [comment]
Would be good to have some test coverage added to prevent another regression reoccurring here as well.  Knowing whether the tests passed or failed is a fairly fundamental requirement, and this is a problem I have to continually work around with the current state of things.

**Comment by gpenghe** [comment]
> The change that originally fixed this problem [2cae4e5#diff-4bca1dda9388536f3fec12ad3707b5f6R370-R373](https://github.com/ThrowTheSwitch/Ceedling/commit/2cae4e55e201ec2f7583e2927430f7a0d16fe1fa#diff-4bca1dda9388536f3fec12ad3707b5f6R370-R373) has been overwritten the very next day: [ebc385f#diff-4bca1dda9388536f3fec12ad3707b5f6L370-L373](https://github.com/ThrowTheSwitch/Ceedling/commit/ebc385f6ff0b30fb0a1ed7e237fb5e3cf3d23593#diff-4bca1dda9388536f3fec12ad3707b5f6L370-L373) Needles to say that restoration of the very last line fixes the problem for me (at least in some cases).

What do you mean "restoration" of the very last line? I am interested to work around this.
